### ISSUE TITLE
[BUG] Plotting ~ 35 metric curves causes significant browser lag

### ISSUE THREAD

**Issue by dbczumar** [mlflow/mlflow#2719]
Thank you for submitting an issue. Please refer to our [issue policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)
for information on what types of issues we address. For help with debugging your code, please refer to [Stack Overflow](https://stackoverflow.com/questions/tagged/mlflow).

Please fill in this template and do not delete it unless you are sure your issue is outside its scope.

### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: OSX 10.14
- **MLflow installed from (source or binary)**: binary
- **MLflow version (run ``mlflow --version``)**: 1.7
- **Python version**: 3.7
- **npm version, if running the dev UI**:
- **Exact command to reproduce**: See steps below

### Describe the problem
Plotting a large number of metrics, each with multiple values recorded at different steps, produces significant browser lag in the UI. i.e., this UI is slow (and pretty unreadable):

<img width="2474" alt="Screen Shot 2020-04-16 at 12 13 09 PM" src="https://user-images.githubusercontent.com/39497902/79496893-acaff700-7fdb-11ea-9a10-3d0d2e630be6.png">

### Code to reproduce issue
1. Launch the MLflow UI / Tracking Server

2. Create a python script with the following contents:
```
import argparse
import time
import random

import numpy as np

import mlflow
from mlflow.tracking.client import MlflowClient


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='PyTorch MNIST Example')
    parser.add_argument('--runs', type=int, default=1, help='The number of runs to generate')
    parser.add_argument('--metrics', type=int, default=5, help='The number of metrics to generate for each run')
    parser.add_argument('--min-step', type=int, default=-10, help='The minimum step to record for each metric')
    parser.add_argument('--max-step', type=int, default=100, help='The maximum step to record for each metric')
    parser.add_argument('--min-value', type=float, default=-10, help='The maximum value to record for each metric')
    parser.add_argument('--max-value', type=float, default=10, help='The maximum value to record for each metric')
    parser.add_argument('--entry-dropout', type=float, default=0.1, help='The fraction of entries to randomly omit')
    parser.add_argument('--step-replication', type=float, default=0.1, help='The fraction of entries that should be given the same step')
    parser.add_argument('--shuffle', action="store_true", help='If specified, shuffles timestamps')

    args = parser.parse_args()

    mlflow_client = MlflowClient()

    for i in range(args.runs):
        with mlflow.start_run():
            run_id = mlflow.active_run().info.run_uuid
            for j in range(args.metrics):
                metric_name = "metric_{idx}".format(idx=j)

                steps = range(args.min_step, args.max_step)
                replicated_steps = np.random.choice(steps, size=int(len(steps) * args.step_replication))
                steps = np.concatenate([steps, replicated_steps])

                curr_time = int(time.time())
                timestamps = [1000 * item for item in range(curr_time, curr_time + len(steps))]
                if args.shuffle:
                    np.random.shuffle(timestamps)

                steps_timestamps = list(zip(steps, timestamps))
                sample_indices = random.sample(
                    range(len(steps_timestamps)),
                    int((1 - args.entry_dropout) * len(steps_timestamps)))
                steps_timestamps = [steps_timestamps[i] for i in sorted(sample_indices)]

                values = (args.max_value - args.min_value) * np.random.random(len(steps_timestamps)) + args.min_value

                for k in range(len(steps_timestamps)):
                    step, timestamp = steps_timestamps[k]
                    value = values[k]
                    print(step, timestamp, value)
                    mlflow_client.log_metric(
                        run_id=run_id,
                        key=metric_name,
                        value=value,
                        timestamp=timestamp,
                        step=step)
```

3. Invoke `python gen_metrics.py --runs 1 --metrics 35 --min-step -10 --max-step 100    --min-value -10 --max-value 10 --entry-dropout 0 --step-replication 0.0` from the command line

4. Open the run metrics plot view and select all 35 metrics that were produced to plot each curve

5. Observe browser lag

### Other info / logs
NA

**Comment by dbczumar** [comment]
cc @Valentyn1997

**Comment by Valentyn1997** [comment]
It is still an open issue for plotly, I guess.

For example, I found only open discussions:
- https://community.plotly.com/t/how-to-efficiently-render-many-70-plots/24829
- https://community.plotly.com/t/page-very-slow-to-load-many-graphs/4310
- https://community.plotly.com/t/plotly-notebook-very-slow-render-with-many-graphs/16861
- https://community.plotly.com/t/how-to-render-many-charts-without-crashing-the-browser/6425

The possible solution I can think of is giving up interactivity and rendering MetricView as a static image.

**Comment by dbczumar** [comment]
@Valentyn1997 Thanks for digging into this! Bummer that plotly performance limitations are a well-known issue. @Zangr any ideas here?

**Comment by leeoniya** [comment]
allow me to shamelessly plug

https://github.com/leeoniya/uPlot

**Comment by Zangr** [comment]
Thank you @dbczumar for filing this, and @Valentyn1997 and @leeoniya for the references, we should definitely keep an eye on the ongoing Plotly performance improvement efforts and perform upgrades accordingly.

**Comment by jeremyjordan** [comment]
+1 would love to see some improvements to the plotting library, it's quite slow to load loss curves for training neural network models.

**Comment by TheMightiestCarrot** [comment]
maybe implementing this package could help resove this issue:
https://github.com/predict-idlab/plotly-resampler
(ill just leave it here and fly away)
